### PR TITLE
Fix: consumer epoch may be inconsistent with the broker

### PR DESCRIPTION
**Pull request by shibd**
### Motivation

Fixes: #14465

In the current implementation [PIP-84](https://github.com/apache/pulsar/wiki/PIP-84-:-Pulsar-client:-Redeliver-command-add-epoch), the  consumer epoch and broker epoch may be inconsistent. Unless reconnected, otherwise consumers will always filter message.

When the consumer calls the `consumer.redeliverUnacknowledgedMessages()` method,  Consumer every time epoch + 1.
But broker may be not increment. In the following code.

https://github.com/apache/pulsar/blob/4f1e39b6921ea401b8c27f17a041d06d85f8abf8/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L310-L322

### Modifications

- When broker receives a message from consumer`redeliver message`,   every time epoch increment.
- Add synchronized to readMoreEntries method.
- Change log print.

### Documentation
- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java [hunk 1]
```diff
@@ -293,6 +293,12 @@ public void redeliverUnacknowledgedMessages(Consumer consumer, long consumerEpoc
     }
 
     private synchronized void internalRedeliverUnacknowledgedMessages(Consumer consumer, long consumerEpoch) {
+
+        if (consumerEpoch > consumer.getConsumerEpoch()) {
+            log.info("set epoch: " + consumerEpoch);
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:298]: only debug level is well, should add topicName, subName and consumerID

Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:298]: Should be a debug level log, otherwise, for each redelivery request, the broker will print log here.
And it's better to provide more information in the log, such as the topic name, the subscription name, the consumer id/name

Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java]: Thanks, fixed.

Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java]: Thanks, fixed.

**Review by codelipenghui — APPROVED** [review]

**Review by congbobo184 — APPROVED** [review]

**Review by nodece — APPROVED** [review]
LGTM

**Review by Jason918 — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java [hunk 2]
```diff
@@ -352,15 +357,17 @@ protected void readMoreEntries(Consumer consumer) {
...
+                havePendingRead = true;
+                if (consumer.readCompacted()) {
+                    topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
+                            this, consumer);
+                } else {
+                    ReadEntriesCtx readEntriesCtx =
+                            ReadEntriesCtx.create(consumer, consumer.getConsumerEpoch());
+                    cursor.asyncReadEntriesOrWait(messagesToRead,
+                            bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
+                }
```
Comment (BewareMyPower) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: Could you explain why did you add the `synchronized` block here? It's redundant because the only modification on this object is `havePendingRead` but it's volatile so the write operation is thread safe. While the following logic doesn't rely on the value of `havePendingRead` and doesn't modify any state of `PersistentDispatcherSingleActiveConsumer` itself.

Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: /cc @Jason918 @codelipenghui @nodece @congbobo184 WDYT?

Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: oh, you are right. I was originally concerned that calling the cursor method is not thread-safe, I looked at the code again, the processing here is originally the way of optimistic locking. Thank your review.

@congbobo184 can you help make sure?

Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: because the readMoreEntries isn't a synchronized method. If any other thread invoke `readMoreEntries` and other thread 
https://github.com/shibd/pulsar/blob/83a983dd21b48a06ef23db8a07b5536d60de8ca8/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L287-L313

it may use the new epoch and use the old readPostion, it will lose message .so add a `synchronized` block

Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: But why should we synchronize the following code block, not the whole `readMoreEntries` or just `calculateToRead` method?

```java
            havePendingRead = true;
            if (consumer.readCompacted()) {
                topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
                        this, consumer);
            } else {
                ReadEntriesCtx readEntriesCtx =
                        ReadEntriesCtx.create(consumer, consumer.getConsumerEpoch());
                cursor.asyncReadEntriesOrWait(messagesToRead,
                        bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
            }
```

Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: > it may use the new epoch and use the old readPostion

I don't get what the `old readPosition` represents.

I mean, why not

```java
    protected synchronized void readMoreEntries(Consumer consumer) {
        /* ... */
    }
```

or

```java
    protected void readMoreEntries(Consumer consumer) {
        // consumer can be null when all consumers are disconnected from broker.
        // so skip reading more entries if currently there is no active consumer.
        if (null == consumer) {
            return;
        }

        synchronized (this) {
            if (consumer.getAvailablePermits() > 0) {
                Pair<Integer, Long> calculateResult = calculateToRead(consumer);
                /* ... */
            }
        }
```

or synchronize other code block, but just the code block I pointed out before:

```java
            havePendingRead = true;
            if (consumer.readCompacted()) {
                topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
                        this, consumer);
            } else {
                ReadEntriesCtx readEntriesCtx =
                        ReadEntriesCtx.create(consumer, consumer.getConsumerEpoch());
                cursor.asyncReadEntriesOrWait(messagesToRead,
                        bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
            }
```

**Please point out which variables need synchronizing**. As far as I see:
- `messagesToRead` and `bytesToRead` are computed by `calculateToRead`, which is not protected by a synchronized block.
- `consumer`, `readEntriesCtx` are related to the `consumer` argument of `readMoreEntries`, and I think these code doesn't change any state of the consumer.
- `CompactTopicImpl#asyncReadEntriesOrWait` is a synchronized method.
-  `ManagedCursorImpl#asyncReadEntriesOrWait` is not a synchronized method but it's thread safe, otherwise there is no need to use an atomic reference (`WAITING_READ_OP_UPDATER`) in it.

Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: The first this PR only fix the new consumer epoch in `OpReadEntry` and OpReadEntry use old readPosition to read `ManagedLedger`.

steps to reproduce:
1. one thread invoke this method
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L296
and then change the cosnumerEpoch from 0 -> 1
this method hasn't rewind the cursor 
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L319
2. one thread invoke readMoreEntry
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L361-L364
now the cursor readPostion hasn't been changed but consumerEpoch has changed. this readOp read the messages, the client will not filter. If you want to redeliver 1-10, but send you the messages is from 10-15 and cumulative ack then client broken down 1-10 messages will loss.

The second is because https://github.com/apache/pulsar/pull/15046 changed the code. 

```
        if (!havePendingRead) {
            cursor.rewind();
            if (log.isDebugEnabled()) {
                log.debug("[{}-{}] Cursor rewinded, redelivering unacknowledged messages. ", name, consumer);
            }
            readMoreEntries(consumer);
        } else {
            log.info("[{}-{}] Ignoring reDeliverUnAcknowledgedMessages: cancelPendingRequest on cursor failed", name,
                    consumer);
        }
```
to

> https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L317-L323

so`havePendingRead` also doesn't need to add the synchronized block. any other code also doesn't need to move into the synchronized block.

```
ManagedCursorImpl#asyncReadEntriesOrWait is not a synchronized method but it's thread safe, otherwise there is no need to use an atomic reference (WAITING_READ_OP_UPDATER) in it.
```
asyncReadEntriesOrWait is thread-safe but it can't solve 
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java#L48 use old readPostion

and 
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L617 use new consumer epoch.

The third, because this is the critical path of reading, we try to reduce the code in the lock as much as possible, other code doesn't need to add to the synchronized block

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: @congbobo184 can we just make `consumer.getConsumerEpoch()`and `consumer.setConsumerEpoch()` keep synchronized?

Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:370]: @codelipenghui  I think this lock is mainly intended for `cursor.rewind()` in `internalRedeliverUnacknowledgedMessages`, which makes the epoch change and cursor rewind atomic.
